src/engine/Service.cpp

```diff
@@ -164,78 +166,109 @@ ProtoResult Service::computeResultImpl([[maybe_unused]] bool requestLaziness) {
+                 parsedServiceClause_.visibleVariables_.end(),
+                 std::back_inserter(expVariableKeys),
+                 [](const Variable& v) { return v.name().substr(1); });
 
   // Set basic properties of the result table.
   IdTable idTable{getResultWidth(), getExecutionContext()->getAllocator()};
   LocalVocab localVocab{};
+
+  auto body = ad_utility::LazyJsonParser::parse(std::move(response.body_),
+                                                {"results", "bindings"});
+
   // Fill the result table using the `writeJsonResult` method below.
-  size_t resWidth = getResultWidth();
-  CALL_FIXED_SIZE(resWidth, &Service::writeJsonResult, this, resVariables,
-                  resBindings, &idTable, &localVocab);
+  CALL_FIXED_SIZE(getResultWidth(), &Service::writeJsonResult, this,
+                  expVariableKeys, body, &idTable, &localVocab);
 
   return {std::move(idTable), resultSortedOn(), std::move(localVocab)};
 }
 
 // ____________________________________________________________________________
 template <size_t I>
 void Service::writeJsonResult(const std::vector<std::string>& vars,
-                              const std::vector<nlohmann::json>& bindings,
+                              cppcoro::generator<nlohmann::json>& body,
                               IdTable* idTablePtr, LocalVocab* localVocab) {
   IdTableStatic<I> idTable = std::move(*idTablePtr).toStatic<I>();
   checkCancellation();
   std::vector<size_t> numLocalVocabPerColumn(idTable.numColumns());
 
-  size_t rowIdx = 0;
-  for (const auto& binding : bindings) {
-    idTable.emplace_back();
-    for (size_t colIdx = 0; colIdx < vars.size(); ++colIdx) {
-      TripleComponent tc = binding.contains(vars[colIdx])
-                               ? bindingToTripleComponent(binding[vars[colIdx]])
-                               : TripleComponent::UNDEF();
-
-      Id id = std::move(tc).toValueId(getIndex().getVocab(), *localVocab);
-      idTable(rowIdx, colIdx) = id;
-      if (id.getDatatype() == Datatype::LocalVocabIndex) {
-        ++numLocalVocabPerColumn[colIdx];
+  // Verifies that the result header contains the expected variables.
+  auto verifyVariables = [&](const nlohmann::json& part) {
+    auto vars = part["head"]["vars"].get<std::vector<std::string>>();
+    ad_utility::HashSet<Variable> responseVars;
+    for (const auto& v : vars) {
+      responseVars.emplace("?" + v);
+    }
+    ad_utility::HashSet<Variable> expectedVars(
+        parsedServiceClause_.visibleVariables_.begin(),
+        parsedServiceClause_.visibleVariables_.end());
+
+    if (responseVars != expectedVars) {
+      throw std::runtime_error(absl::StrCat(
+          "Header row of JSON result for SERVICE query is \"",
+          absl::StrCat("?", absl::StrJoin(vars, " ?")), "\", but expected \"",
+          absl::StrJoin(parsedServiceClause_.visibleVariables_, " ",
+                        Variable::AbslFormatter),
+          "\"."));
+    }
+  };
```

Comment:
Can this be a separate function?
It is rather much code just for a sanity check.
Maybe a function that takes 
* a partial json
* the visible variables
And throws on error.

Do I remember correctly, that the lazy json parser gives us the head in exactly one batch, so the check `if part.contains("head")` will be true exactly once? Otherwise we could omit all but the first redundant check.